pkg/retry/retry_test.go

```diff
@@ -121,3 +126,96 @@ func (s *runSuite) TestInfiniteRetry(c *check.C) {
 	c.Assert(reportedElapsed, check.Greater, time.Second)
 	c.Assert(reportedElapsed, check.LessEqual, 3*time.Second)
 }
+
+func (s *runSuite) Test_DoShouldRetryAtMostSpecifiedTimes(c *check.C) {
+	defer testleak.AfterTest(c)()
+	var callCount int
+	f := func() error {
+		callCount++
+		return errors.New("test")
+	}
+
+	err := Do(context.Background(), f, WithMaxTries(3))
+	c.Assert(err, check.ErrorMatches, "*test")
+	c.Assert(callCount, check.Equals, 3)
+}
+
+func (s *runSuite) Test_DoShouldStopOnSuccess(c *check.C) {
+	defer testleak.AfterTest(c)()
+	var callCount int
+	f := func() error {
+		callCount++
+		if callCount == 2 {
+			return nil
+		}
+		return errors.New("test")
+	}
+
+	err := Do(context.Background(), f, WithMaxTries(3))
+	c.Assert(err, check.IsNil)
+	c.Assert(callCount, check.Equals, 2)
+}
+
+func (s *runSuite) Test_IsRetryable(c *check.C) {
+	defer testleak.AfterTest(c)()
+	var callCount int
+	f := func() error {
+		callCount++
+		return errors.Annotate(context.Canceled, "test")
+	}
+
+	err := Do(context.Background(), f, WithMaxTries(3), WithIsRetryableErr(func(err error) bool {
+		switch errors.Cause(err) {
+		case context.Canceled:
+			return false
+		}
+		return true
+	}))
+
+	c.Assert(errors.Cause(err), check.Equals, context.Canceled)
+	c.Assert(callCount, check.Equals, 1)
+
+	callCount = 0
+	err = Do(context.Background(), f, WithMaxTries(3))
+
+	c.Assert(errors.Cause(err), check.Equals, context.Canceled)
+	c.Assert(callCount, check.Equals, 3)
+}
+
+func (s *runSuite) Test_DoCancelInfiniteRetry(c *check.C) {
+	defer testleak.AfterTest(c)()
+	callCount := 0
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*20)
+	defer cancel()
+	f := func() error {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+		callCount++
+		return errors.New("test")
+	}
+
+	err := Do(ctx, f, WithInfiniteTries(), WithBackoffBaseDelay(2), WithBackoffMaxDelay(10))
+	log.Info("Test_DoCancelInfiniteRetry run f times", zap.Int("tries", callCount))
```

Comment:
It's better to delete the log in the unit tests and use `Commentf` if you want print some extra information.